token/js/src/actions/approve.ts

```diff
@@ -1,15 +1,8 @@
-import {
-    ConfirmOptions,
-    Connection,
-    PublicKey,
-    sendAndConfirmTransaction,
-    Signer,
-    Transaction,
-    TransactionSignature,
-} from '@solana/web3.js';
-import { TOKEN_PROGRAM_ID } from '../constants';
-import { createApproveInstruction } from '../instructions/index';
-import { getSigners } from './internal';
+import type { ConfirmOptions, Connection, PublicKey, Signer, TransactionSignature } from '@solana/web3.js';
+import { sendAndConfirmTransaction, Transaction } from '@solana/web3.js';
+import { TOKEN_PROGRAM_ID } from '../constants.js';
+import { createApproveInstruction } from '../instructions/approve.js';
+import { getSigners } from './internal.js';
```

Comment:
What's changing across the project is that we are using imports with explicit .js extensions, and we're limiting use of index imports to descendants of any given directory. Each instruction can be imported directly without circular references.

Reply:
what's the reason for explicit .js extensions? Is that specifically for those build systems?  And what do you mean by

> we're limiting use of index imports to descendants of any given directory

I didn't notice any differences in the import statements in index files

Reply:
> what's the reason for explicit .js extensions? Is that specifically for those build systems?

TypeScript (somewhat infamously) [will not add `.js` extensions to imports](https://github.com/microsoft/TypeScript/issues/16577) when compiling, but _will_ resolve `.js` files to `.ts` files.

Next.js, Webpack 5, and some other build systems won't automatically resolve `.mjs` files when importing without an extension, which is what is causing https://github.com/solana-labs/solana-program-library/issues/3502.

Explicit declaring the file extensions here is the way to get them explicitly declared in the output. I will probably need to add an eslint rule to enforce this.

> I didn't notice any differences in the import statements in index files

For example, this file change we are commenting on replaces:

```ts
import { createApproveInstruction } from '../instructions/index';
```
with 
```ts
import { createApproveInstruction } from '../instructions/approve.js';
```

Instead of using an index import, which would need to become `'../instructions/index.js'` anyway, we just import from the specific file we need. This practice is generally good because it helps avoid circular references. Plus, we don't need to resolve the whole index to get one or two exports from it.

Reply:
Ah gotcha, I missed that part, thanks for all the explanation!

Reply:
Np!